### PR TITLE
Replace Array#| with Array#+

### PR DESCRIPTION
**Pull request by seancaffery**
This is primarily for performance reasons. It yields around a 23% improvement on a file with 300 violations.

Here is a benchmark using [readygo](https://github.com/garybernhardt/readygo):
```
lint check ticket
  Baseline: |                                                X-                |
  Current:  |                                                                X-|
            0                                                           13.288 s
```
Baseline (with this change) is around 9.6s.

From my understanding this will not have an impact on functionality even though there can be duplicates in `excludes`. There may be a performance impact on the `include?` call but the improvement seems to be worth it.

### CODE REVIEW

**Comment by YorickPeterse** [comment]
I can't really grok the output of the benchmark sadly. Could you do me a favour and write one using https://github.com/evanphx/benchmark-ips? This Gem takes care of warming up things (which can make a big difference when a JIT is involved) which typically leads to more reliable results. Also, please share the benchmark code so I can double check things.

Functionality wise this looks fine to me.

**Comment by seancaffery** [comment]
That's fair enough. I did run them the wrong way around.
Readygo accounts for things like GC and jitter. If you're interested, you can take a look at the [timing methodology](https://github.com/garybernhardt/readygo#timing-methodology).

I did the benchmark with benchmark-ips. Here are the results.
**master**
```
Calculating -------------------------------------
             or vs +     1.000  i/100ms
-------------------------------------------------
             or vs +      0.077  (± 0.0%) i/s -      1.000  in  12.932026s
```

**my branch**
```
Calculating -------------------------------------
             or vs +     1.000  i/100ms
-------------------------------------------------
             or vs +      0.104  (± 0.0%) i/s -      1.000  in   9.576817s
```

bench.rb
```ruby
require_relative 'lib/ruby-lint'
require 'benchmark/ips'

Benchmark.ips do |x|
  x.report('or vs +') do
    `ruby bin/ruby-lint model.rb`
  end
end

```
This isn't ideal, but there is something in ruby-lint that messes with the output of benchmark-ips and I don't know enough yet to understand what it is.

**Comment by YorickPeterse** [comment]
So I did some digging again to see why this was using `Array#|` in the first place. Basically this is to ensure that `self` is not added multiple times, this was added in 08c80f60a193014e430ee1b64e88c1e2ea9ca82f. By using `Array#+` this check is lost, meaning it could in theory be possible for `self` to end up multiple times in the array.

To check the performance _without_ involving the entire CLI I wrote the following benchmark:

```ruby
require 'benchmark/ips'
require_relative 'lib/ruby-lint'

definitions = RubyLint::Definition::RubyObject.new

RubyLint.registry.load('Encoding')
RubyLint.registry.apply('Encoding', definitions)

encoding = definitions.lookup(:const, 'Encoding')

Benchmark.ips do |bench|
  bench.report 'RubyObject#lookup' do
    encoding.lookup(:const, 'Kernel')
  end
end
```

This requires some insight in how ruby-lint works, so don't worry if this doesn't make a lot of sense. Basically it loads a definition for `Encoding` and then looks up something defined somewhere higher in the ancestor chain (`Kernel` in this case), triggering the code in question.

On my laptop the output without any changes is as following:

    Calculating -------------------------------------
       RubyObject#lookup    21.711k i/100ms
    -------------------------------------------------
       RubyObject#lookup    278.391k (± 1.1%) i/s -      1.411M

By switching to `Array#+` the output changes to the following:

    Calculating -------------------------------------
       RubyObject#lookup    27.376k i/100ms
    -------------------------------------------------
       RubyObject#lookup    378.418k (± 1.2%) i/s -      1.916M

To ensure the `self` check is still in place I used the following patch instead:

```diff
diff --git a/lib/ruby-lint/definition/ruby_object.rb b/lib/ruby-lint/definition/ruby_object.rb
index 1193674..c4a045e 100644
--- a/lib/ruby-lint/definition/ruby_object.rb
+++ b/lib/ruby-lint/definition/ruby_object.rb
@@ -697,7 +697,9 @@ module RubyLint
         if parent.type == type and parent.name == name
           parent_definition = parent
         else
-          parent_definition = parent.lookup(type, name, true, exclude | [self])
+          exclude = exclude + [self] unless exclude.include?(self)
+
+          parent_definition = parent.lookup(type, name, true, exclude)
         end

         return parent_definition
```

If we then re-run the benchmark the output is as following:

    Calculating -------------------------------------
       RubyObject#lookup    25.734k i/100ms
    -------------------------------------------------
       RubyObject#lookup    338.675k (± 2.8%) i/s -      1.698M

To summarize:

    Array#|            278.391k (± 1.1%) i/s -      1.411M
    Array#+            378.418k (± 1.2%) i/s -      1.916M
    Array#+ with check 338.675k (± 2.8%) i/s -      1.698M

Personally I'm satisfied with the combination of `Array#+` and an
`Array#include?` check as it retains the guard while still speeding up things a
bit. If you'd like you can apply the patch above to your pull-request, otherwise
I'll commit it myself.

Out of curiosity, did you find this specific bit of code by using some sort of
profiling tool, or did you end up at it based on a gut feeling of sorts?

**Comment by seancaffery** [comment]
I wasn't sure if it mattered about the duplicates, but it seems you put it there for a reason. I updated my branch to include your patch.

Thanks for the explanation on how to profile this. It will make future testing a lot easier and faster.

I just used ruby-prof. I thought it was a bit strange that `Array#|` kept appearing so I went and looked at what it was doing and replaced it with something similar.
The drive to start looking at this was having synstastic use ruby-lint as a checker, but it was taking too long to open files in my project. I think there are other wins to be had, but I need to spend some more time understanding the code before I have a go at those. The code documentation definitely makes this a lot easier :)
